Add nekRS package

thilinarmtb opened the pull request.

thilinarmtb committed:
Add nekRS package

thilinarmtb committed:
Remove the patch retained from Nek5000 package

thilinarmtb committed:
Use spack Executable to run `makenrs`

thilinarmtb committed:
Fix vermin failure

thilinarmtb committed:
Fix a style error

sethrj commented on var/spack/repos/builtin/packages/nekrs/package.py:
```diff
@@ -0,0 +1,189 @@
+
+        fflags = ' '.join(fflags)
+        cflags = ' '.join(cflags)
+        cxxflags = ' '.join(cxxflags)
```
Your flags don't seem to be used as part of the actual build? I think 154 through 173 could justs be removed.

thilinarmtb replied:
I removed them from the package.

sethrj commented on var/spack/repos/builtin/packages/nekrs/package.py:
```diff
@@ -0,0 +1,189 @@
+        else:
+            env.set('OCCA_HIP_ENABLED', '0')
+
+        if '+opencl' in spec:
+            env.set('OCCA_OPENCL_ENABLED', '1')
+        else:
+            env.set('OCCA_OPENCL_ENABLED', '0')
```
```suggestion
        env.set('OCCA_OPENCL_ENABLED', if '+opencl' in spec else '0')
```
and elsewhere above

sethrj reviewed (changes requested):
A few comments about the recipe -- plus a major question: did you look at subclassing `CudaPackage` and `ROCmPackage`? I don't see anywhere (for example) that the CUDA target architectures are defined.

thilinarmtb committed:
Remove unused code

thilinarmtb committed:
Update var/spack/repos/builtin/packages/nekrs/package.py

Co-authored-by: Seth R. Johnson <johnsonsr@ornl.gov>

thilinarmtb commented:
@sethrj : Thanks for the review ! I fixed some of the issues. I am not sure I understand your comment about `CudaPackage` and `ROCmPackage`. Can you elaborate a bit more?

thilinarmtb committed:
Fix a syntax error

sethrj commented on var/spack/repos/builtin/packages/nekrs/package.py:
```diff
@@ -0,0 +1,167 @@
+        makenrs.add_default_env("NEKRS_INSTALL_DIR", prefix)
+        makenrs.add_default_env("NEKRS_CC", cc)
+        makenrs.add_default_env("NEKRS_CXX", cxx)
+        makenrs.add_default_env("NEKRS_FC", fc)
```
Since the variables are now used only here I recommend putting them on the same line.

sethrj commented on var/spack/repos/builtin/packages/nekrs/package.py:
```diff
@@ -0,0 +1,167 @@
+        else:
+            env.set('OCCA_HIP_ENABLED', '0')
+
+        env.set('OCCA_OPENCL_ENABLED', if '+opencl' in spec else '0')
```
I think you mean
```suggestion
        env.set('OCCA_OPENCL_ENABLED', '1' if '+opencl' in spec else '0')
```

sethrj commented on var/spack/repos/builtin/packages/nekrs/package.py:
```diff
@@ -0,0 +1,167 @@
+        if '+hip' in spec:
+            env.set('OCCA_HIP_ENABLED', '1')
+        else:
+            env.set('OCCA_HIP_ENABLED', '0')
```
```suggestion
       env.set('OCCA_HIP_ENABLED', '1' if '+hip' in spec else '0')
```

sethrj commented:
Sure. Is this your first Spack package with CUDA/HIP support? I recommend reading the documentation at  https://spack.readthedocs.io/en/latest/build_systems/rocmpackage.html and https://spack.readthedocs.io/en/latest/build_systems/cudapackage.html .

You'll probably want to make the Nekrs package inherit from both CUDA and ROCm as mixins:
```python
class Nekrs(Package, CudaPackage, ROCmPackage):
```
and then instead of using `+hip` you'll use `+rocm`.

Here's an example of a package that uses both HIP and CUDA, although it's through cmake: https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/camp/package.py

thilinarmtb commented:
@sethrj : Thanks for the suggestion ! Yes, this is my first package with `HIP` and/or `CUDA` and I was following `occa` package as it uses `CUDA` and `HIP`. Let me have a look at what you suggested.

thilinarmtb committed:
Minor refactoring

sethrj commented:
Great! Ok I'm not too familiar with OCCA nor have I ever built nekrs. It sounds like you haven't tested the spack installation of nekrs then?

sethrj commented:
@thilinarmtb I looked into this a bit more and if nekRS is using OCCA, then I think you need to get rid of the environment variables you've set, and variants, and just use `depends_on('occa')`. Perhaps one of the spack OCCA maintainers v-@dobrev @dmed256 could help you out?

thilinarmtb commented:
@sethrj : I tested the package as is on `ORNL Summit` and installs fine. We do use `OCCA` but we use a custom version of it (not the one shipped with `spack`).

sethrj commented:
> We do use `OCCA` but we use a custom version of it (not the one shipped with `spack`).

Oh, well if you're using a custom forked version of `OCCA` then your current package might be the right approach... but to be a good citizen of the larger scientific software ecosystem, your group should think seriously about allowing OCCA (and any other "built-in" libraries your project has) to be an external dependency -- otherwise if someone wants to make a coupled CFD/whatever code and the "whatever" code also requires OCCA or another dependency, that coupled code will break. This does actually happen in real life and in fact can be the source of nasty bugs due to ODR violations. But this is a longer-term issue so for now basing `nekRS` off the `OCCA`  package is probably the best approach.

chuckatkins commented:
> Sure. Is this your first Spack package with CUDA/HIP support? I recommend reading the documentation at https://spack.readthedocs.io/en/latest/build_systems/rocmpackage.html and https://spack.readthedocs.io/en/latest/build_systems/cudapackage.html .
> 
> You'll probably want to make the Nekrs package inherit from both CUDA and ROCm as mixins:
> 
> ```python
> class Nekrs(Package, CudaPackage, ROCmPackage):
> ```
> 
> and then instead of using `+hip` you'll use `+rocm`.
> 
> Here's an example of a package that uses both HIP and CUDA, although it's through cmake: https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/camp/package.py

nekRS is also a CMake package.  The `makenrs` build script is just a shell script that calls CMake based on environment variables.  The purpose of it seems to be to abstract the build and packaging away from the user.  In this case `spack` takes the place of whatever `makenrs` does.  I think it would be much more preferable to have the spack package directly be a `CMakePackage`, i.e.
```python
class Nekrs(CMakePackage, CudaPackage, ROCmPackage):
```

thilinarmtb commented:
@chuckatkins Thanks for the comments !  `makenrs` changes a lot (especially since `nekrs` is going through frequent and backwards incompatible changes) and replicating it in `spack`  will create maintenance trouble since we will need to change `spack` package whenever the `makenrs` changes. So I prefer the `spack` package to just call `makenrs`.

thilinarmtb committed:
Inherit from Cuda and ROCm

and removed develop version as it is not stable.

thilinarmtb commented:
@sethrj : I inherited from `CudaPackage` and `ROCmPackage` as you requested.

thilinarmtb committed:
Remove post installation check

as it doubles the installation time (~5mins) and fails on
Summit unless the `-pami_noib` flag is passed to mpirun.

sethrj commented:
@thilinarmtb I strongly advise you to listen to @chuckatkins on this. One major headache is that (with the current version of Spack) if you choose to change your build system later in a spack package's life you'll have to make a new spack package (`nekrs-new` or something).

I took a quick look at `makenrs` and it's not very complicated.  It will definitely help spack users of nek if the configure process is more transparent, and since spack can account for build systems that your script hasn't anticipated, it will be more robust and stable in the long run.

Please switch to using the CMake/CUDA/Hip package and I can point you in the direction of an example if you need one.

Thanks!

stgeke commented:
Any chance to merge this PR today? We need this for the upcoming CEED 4.0 release.

sethrj commented:
> Any chance to merge this PR today? We need this for the upcoming CEED 4.0 release.

I think that if the review feedback can get addressed today and the updated recipe tested, there is a chance yes. If you think the review process needs to be overridden for your unanticipated deadline, then you can always ping Todd.

stgeke commented:
I guess the only thing we haven't addressed yet is the usage of cmake instead of our build script. Do you consider this as a MUST or can we change this later on?

sethrj commented:
> I guess the only thing we haven't addressed yet is the usage of cmake instead of our build script. Do you consider this as a MUST or can we change this later on?

Well, I guess the intermediate build system can slide as a stopgap measure... the variants are all going to change when you switch to cmake, and it's likely to be lacking in robustness in the meantime. One final change to make, though, is to please add conflict statements for all the packages that nekRS forcibly builds into its library:
```py
    for pkg in ['occa', 'hyper', 'nek5000', 'blas', 'lapack', ...]:
        conflicts('^' + pkg, msg=(pkg + " is built into nekRS"))
```

and I recommend going forward to find a way to allow nekRS to build against those as external dependencies -- without the conflicts, downstream libraries may see conflicting symbols between nekRS's builtins and the intended libraries needed for other packages.

stgeke commented:
nekRS will typically run as a stand-alone solver, so there are no downstream dependencies.

sethrj commented:
@stgeke I work with the ExaSMR project which has the ENRICO coupled CFD/neutronics solver, and I believe that is using Nek5000 and/or NekRS as an external library rather than an app.

stgeke commented:
That’s why I said “typically”. The issue you bring up is correct but we won’t be able to address it now.

sethrj commented:
@stgeke I'm a spack maintainer trying to make the HPC community a more cohesive, easier-to-work-with ecosystem, volunteering my time to help people integrate packages like yours. Out of the four messages you've posted, all four have been demanding, petulant, and hostile; and I do not appreciate the attitude.

I'm happy to continue working with @thilinarmtb on meeting some minimum requirements for integrating into this ecosystem, and in fact I've not only spelled out what that minimum is, I've gone so far as to *provide the code* to do so. So please try to work with me and recognize that I'm trying to help you out and not trying to antagonize you.

thilinarmtb commented:
@sethrj : Thanks for the input so far. Would adding conflicts will stop spack from installing `nekrs` if `occa` is already installed or vice versa? I tried the code you suggested for adding conflicts, but I was able to do a `spack install nekrs` after doing `spack install occa`. Basically, I did not see any difference as per the user's perspective.

sethrj commented:
@thilinarmtb They can definitely be installed side-by-side. It should simply prevent a package from unintentionally depending on both occa and nekrs -- unlikely to happen if they're working closely with you guys, but good documentation and a good way of being  a good neighbor by declaring your known incompatibilities.

thilinarmtb commented:
Okay, cool ! Thanks. Let me push the changes regarding the conflicts you asked for.

jedbrown commented:
Hmm, it sounds like this conflict will prevent the meta-package `ceed` from depending on both NekRS and OCCA, which it's supposed to do since both are part of the distribution. Can the internal OCCA used by NekRS be namespaced if you can't use a version of upstream OCCA?

thilinarmtb commented:
@jedbrown : Hmm, let me get back to you on that one.

sethrj commented:
OK, thanks @jedbrown for the context -- is there a plan to move toward an E4S (spack.yaml environment) style approach? Packages really will get into trouble if they do try to link against libraries that have these duplicate symbols.

If that's the case, then for now @thilinarmtb can you add the conflict lines but comment them out, and add an explanatory note saying that `nekrs` includes these packages, but in order to build as part of `CEED` we can't leave them in?

jedbrown commented:
@sethrj For `ceed`? It has been a metapackage within spack; WIP [update for 4.0 here](https://github.com/spack/spack/tree/packages/ceed-4.0), with a todo to add NekRS once this current PR merges. I see the [E4S repository](https://github.com/E4S-Project/e4s), but don't know what advantage this would provide. It seems less flexible, but I don't know the rationale.

The CEED metapackage is used by some to explore examples/miniapps usisg software created by the CEED project, but would probably never all be linked at once. Most likely, `mfem` uses `libceed` and `occa` (`libceed` can also link `occa`), and an application can link to any subset of those. There are `libceed` examples using `occa` that are callable using Nek5000, but not currently using NekRS. I don't love the unsound dependency graph we're proposing with NekRS using a special non-namespaced OCCA, but it isn't likely to cause actual user conflicts at this time. I do consider it important to address in the future, however.

thilinarmtb committed:
Add a note about potential conflicts

Currently commented out since we need nekrs to be used in CEED.

thilinarmtb commented:
@sethrj : I added a list of potential conflicts and commented it out. @stgeke: Are there any conflicts I missed [here](https://github.com/spack/spack/pull/22519/commits/111330461127e813175808887ac215f2a3d8110d#diff-38806bdde3b783bf9bb54f30e3351dc29cf22c747ae86c97e684c1fe94f4d5c9R39) ?

stgeke commented:
@thilinarmtb Not sure if it applies but how about parRSB and gslib?

sethrj reviewed (approved):
I'll approve as-is since the people on this PR discussion seem to be the only users for now -- but please add a to-do item to convert this to a standard `CMakePackage`. Currently none of the CUDA/HIP options from `CudaPackage` or `ROCmPackage` are actually being used by the recipe, and `+hip` should probably replaced by `+rocm` here.

I also strongly recommend adding an issue on your nekrs tracker to support external installations of your dependencies. Contrary to what I heard in this discussion, it looks like other users have *already* encountered conflicts due to the built-in NekRS libraries  https://github.com/Nek5000/nekRS/issues/180 .

Since I hear that NekRS is impressively fast and is rapidly developing capabilities, the number of interested users is likely to increase -- and having a one-line spack install for it is going to make it easier for apps to want to try it out as a library. So in the interests of lowering your future maintenance burden, especially knowing from experience that it's very easy in CMake to support external library installations while retaining an internal option for convenience, I hope you  can quickly improve your ability to integrate into a larger software ecosystem.

Welcome to spack!

thilinarmtb commented:
Thanks for all the help @sethrj !
